### PR TITLE
Support async collective op execution

### PR DESCRIPTION
**Pull request by eedalong**

### CODE REVIEW

File: tao_compiler/mlir/disc/disc_compiler.cc [hunk 1]
```diff
@@ -529,6 +530,14 @@ LogicalResult LowerHLOToLLVM(ModuleOp m, const DISCLoweringOptions& options) {
   pm.addNestedPass<FuncOp>(createCanonicalizerPass());
   pm.addNestedPass<FuncOp>(createCSEPass());
   pm.addNestedPass<FuncOp>(createCanonicalizerPass());
+
+  bool enable_op_schedule = false;
+  tensorflow::ReadBoolFromEnvVar("DISC_ENABLE_OP_SCHEDULE", enable_op_schedule,
```
Comment (Yancey1989) [inline tao_compiler/mlir/disc/disc_compiler.cc:535]: Does this pass cause any native optimization? If not, let's enable it by default, we have too many flags...

File: tao_compiler/mlir/disc/transforms/disc_collective_ops_rewriter.cc [hunk 2]
```diff
@@ -55,6 +55,24 @@ std::optional<std::string> ReductionKindToString(ReductionKind kind) {
   return std::nullopt;
 }
 
+bool EnableAsyncCollective(Operation* op) {
+  if (llvm::isa<mhlo::AllReduceOp>(op)) {
+    if (const char* env_p = std::getenv("ENABLE_ASYNC_ALL_REDUCE")) {
```
Comment (Yancey1989) [inline tao_compiler/mlir/disc/transforms/disc_collective_ops_rewriter.cc:60]: make all collective ops to be async? Do we need to separate sync and async collective ops ?

File: tao_compiler/mlir/disc/transforms/disc_op_schedule.cc [hunk 3]
```diff
@@ -0,0 +1,992 @@
...
+  return static_cast<int64_t>(resource_usage_type);
+}
+
+using ResourcePair = std::pair<int64_t, ResourceUsageType>;
+using ResourcesVector = std::vector<ResourcePair>;
+using TimeCost = double;
+
+class Edge;
+
+struct GraphNode {
```
Comment (Yancey1989) [inline tao_compiler/mlir/disc/transforms/disc_op_schedule.cc:101]: there is latency hidden schedule in tensorflow source code, can we reuse the base class implementation? 
ref: https://github.com/pai-disc/tensorflow/blob/features/bladedisc_rebase_20230202/tensorflow/compiler/xla/service/latency_hiding_scheduler.h

File: tao_compiler/mlir/disc/transforms/disc_collective_ops_rewriter.cc (hunk 2 above)
Comment (eedalong) [reply tao_compiler/mlir/disc/transforms/disc_collective_ops_rewriter.cc:60]: Just mimicking [openXLA's setting and behaviour](https://github.com/openxla/xla/blob/4c566c945ad38e69bc71b5ee67b741551187e011/xla/xla.proto#L388)

File: tao_compiler/mlir/disc/disc_compiler.cc (hunk 1 above)
Comment (eedalong) [reply tao_compiler/mlir/disc/disc_compiler.cc:535]: Ok, we havent found a case which schedule pass will cause negative effect.

Comment (eedalong) [reply tao_compiler/mlir/disc/disc_compiler.cc]: Fixed.

File: tao_compiler/mlir/disc/transforms/disc_collective_ops_rewriter.cc (hunk 2 above)
Comment (eedalong) [reply tao_compiler/mlir/disc/transforms/disc_collective_ops_rewriter.cc]: Fixed.

File: tao_compiler/mlir/disc/transforms/disc_op_schedule.cc [hunk 4]
```diff
@@ -0,0 +1,999 @@
...
+    }
+
+    scheduler_core_ = new SchedulerCore(latency_estimator_, async_tracker_,
+                                        scheduler_config_);
+    auto scheduled_op_sequence =
+        scheduler_core_->ScheduleComputation(original_op_sequence);
+
+    for (auto& block : main_func.getBody()) {
+      for (int op_idx = 0; op_idx < scheduled_op_sequence.size(); op_idx++) {
+        scheduled_op_sequence[op_idx]->moveBefore(&block.front());
```
[the remark this answers is not in the document]
Comment (eedalong) [reply tao_compiler/mlir/disc/transforms/disc_op_schedule.cc:985]: Yes, it contains all ops.

File: tao_compiler/mlir/disc/transforms/disc_op_schedule.cc [hunk 5]
```diff
@@ -0,0 +1,999 @@
...
+              Edge(user_node,
+                   latency_estimator->GetLatencyBetween(node, user_node)));
+          user_node->predecessors.emplace_back(Edge(
+              node, latency_estimator->GetLatencyBetween(node, user_node)));
+        }
+      }
+    }
+  }
+
+  void InitilizeGrpahTopology(std::vector<Operation*>& post_order_instructions,
```
[the remark this answers is not in the document]
Comment (eedalong) [reply tao_compiler/mlir/disc/transforms/disc_op_schedule.cc:500]: Ok, will fixed it in next pr.

**Pull request merged**